### PR TITLE
Improve IO handling

### PR DESCRIPTION
**Pull request by jkonecny12**
This is major rework of the input processing and input reaction. It shouldn't change API of the SImpleline.

Summary of this change:
* Split input to multiple classes. 
* Input processing is mostly done in `UIScreen` now and every `UIScreen` has its own `InputManager` instance.
* Add `nputHandlers -- objects which creating abstractions to get user input.
* All InputHandlers are using one instance of `InputThreadManager` which creates and runs threads to get input and send this input to the correct `InputHandler`. It also solves concurrent inputs. Still it is not recommended to do ask for multiple inputs at once.
* Source of the input is passed to the `InputHandler` so it is attached to the event processing used by the `UIScreen`.
* If application crash because of concurrent input check then we will get information what InputHandlers asked and what are their requesters (if specified).
* Width is specified for the whole application by adding `App.width` attribute.

--------------------------------------------------------------------------------------

**NOTE:**
Personally I don't like solution with the `InputThreadManager` because this way it is not restarted when `App.initialize()` is called. I welcome any suggestions how to transform `InputThreadManager` to be able to control input thread but also be restarted by `App.initialize()` call.

I was thinking about having callbacks registered to the `App.initialize()`. I guess that would work but I don't like that solution... Do you have better idea?

### CODE REVIEW

File: simpleline/input/__init__.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                ret_list.append(t)
+
+        return ret_list
+
+
+# create singleton of InputThreadManager -- this will be used by InputHandler classes
+inputThreadMgr = InputThreadManager()
+
+
+class InputThread(object, metaclass=ABCMeta):
```
[the remark this answers is not in the document]
Comment (poncovka) [reply simpleline/input/__init__.py]: Isn't this class more like an input request then an input thread?

File: simpleline/render/screen/__init__.py [hunk 2]
```diff
@@ -159,16 +181,25 @@ def window(self, window):
...
+        if hidden:
+            handler = PasswordInputHandler(source=self)
+            if self.password_func:
+                handler.set_pass_func(self.password_func)
+        else:
+            handler = InputHandler(source=self)
+
+        handler.get_input(message)
+        handler.wait_on_input()
+        return handler.value
```
Comment (poncovka) [inline simpleline/render/screen/__init__.py]: I am a little confused. Why the `InputManager` doesn't do this?

File: simpleline/render/screen/__init__.py [hunk 3]
```diff
@@ -228,12 +259,18 @@ def _print_widget(self, widget):
...
                 custom_prompt = Prompt(_("\nPress %s to continue") % Prompt.ENTER)
-                App.get_scheduler().io_manager.get_user_input(custom_prompt)
+                self._ask_user_input_blocking(custom_prompt)
                 pos += real_screen_height
 
+    def _ask_user_input_blocking(self, prompt):
+        handler = InputHandler(source=self)
+        handler.get_input(prompt)
+        handler.wait_on_input()
+        return handler.value
```
Comment (poncovka) [inline simpleline/render/screen/__init__.py]: This could be also done by the `InputManager`.

**Comment by jkonecny12** [comment]
Change behavior on concurrent input. Last requester will receive input others are dropped as failed. This failed state is especially useful when the `wait_on_input` method is blocking thread.

File: simpleline/input/__init__.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+        :param input_thread_object: Input thread object based on InputThread class.
+        :param concurrent_check: Should the concurrent thread check be fatal? (default True).
+        """
+        if self._registered_loop is not App.get_event_loop():
+            App.get_event_loop().register_signal_handler(InputReceivedSignal,
+                                                         self._input_received_handler)
+            self._registered_loop = App.get_event_loop()
+
+        self._check_input_thread_running(concurrent_check, input_thread_object)
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: Good point.

File: simpleline/input/input_handler.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        :returns: User input.
+        :rtype: str
+        """
+        self._clear_input()
+        self._invoke_input_thread(prompt)
+
+    def _invoke_input_thread(self, prompt):
+        thread_object = self.create_thread_object(prompt)
+
+        inputThreadMgr.start_input_thread(thread_object, not self._skip_concurrency_check)
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/input_handler.py]: I think better place would be something like `get_instance` method inside the manager. The point is that `InputHandler` shouldn't care about creation of manager it is not its responsibility.

File: simpleline/input/__init__.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                running_threads = []
+                running_threads.extend(self._input_stack)
+                running_threads.append(new_thread_object)
+                for t in running_threads:
+                    t_source = t.source
+                    if t_source.source:
+                        msg += "Input handler: {} Input requester: {}\n".format(t_source,
+                                                                                t_source.source)
+                    else:
+                        msg += "Input handler: {} Input requester: Unknown\n".format(t_source)
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: Nice improvement, thanks!

File: simpleline/input/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: Documented and renamed.

File: simpleline/input/input_handler.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/input_handler.py]: Fixed.

**Comment by jkonecny12** [comment]
> As this changes the input handling code quite a bit we should definitely test this with Initial Setup and it's multi TTY support to make sure everything works correctly.

I don't think it should change behavior but we can't go wrong with additional testing.

Updated but it is still WIP. Some of the notes are implemented now.

File: simpleline/input/__init__.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    def _get_request_source(self, thread_object):
+        """Get user input request source.
+
+        That means object who is using InputHandler.
+        If this object is not specified then return InputHandler as a source.
+        """
+        if thread_object.source.source:
+            return thread_object.source.source
+        else:
+            return thread_object.source
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: I was thinking about that too. I'll implement it.

File: simpleline/input/__init__.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        super().__init__()
+        self._input_stack = []
+        self._registered_loop = None
+        self._processing_input = False
+
+    def _input_received_handler(self, signal, args):
+        thread_object = self._input_stack[-1]
+        handler_source = thread_object.source
+        signal_source = self._get_request_source(thread_object)
+        App.get_event_loop().enqueue_signal(InputReadySignal(source=signal_source,
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: Fixed.

File: simpleline/input/__init__.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+    def __init__(self, source):
+        super().__init__()
+        self.source = source
+        self.thread = None
+
+    def run(self):
+        data = self.run_input()
+
+        App.get_event_loop().enqueue_signal(InputReceivedSignal(self, data))
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: Fixed.

File: simpleline/input/__init__.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+    @abstractmethod
+    def run_input(self):
+        """Get input from user.
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: Fixed.

File: simpleline/input/__init__.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+from simpleline.logging import get_simpleline_logger
+from simpleline.event_loop.signals import InputReceivedSignal, InputReadySignal
+
+log = get_simpleline_logger()
+
+
+INPUT_THREAD_NAME = "SimplelineInputThread"
+
+
+class InputThreadManager(object):
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: Fixed.

File: simpleline/input/__init__.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+    def _start_user_input_async(self):
+        if not self._input_stack or self._processing_input:
+            return
+
+        thread_object = self._input_stack[-1]
+
+        thread = threading.Thread(name=INPUT_THREAD_NAME, target=thread_object.run)
+        thread.daemon = True
+        thread_object.thread = thread
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: True it is better for SRP.

File: simpleline/input/__init__.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    __instance = None
+
+    def __init__(self):
+        super().__init__()
+        self._input_stack = []
+        self._registered_loop = None
+        self._processing_input = False
+
+    def _input_received_handler(self, signal, args):
+        thread_object = self._input_stack[-1]
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply simpleline/input/__init__.py]: I can implement `emit_input_ready` and `emit_input_failed` signals. I like that solution. 

However, I don't want to directly call input handler. Reason for this solution (even that it is more complex) is a signal source. Point is that I can now specify signal source for `InputReadySignal` of a screen. Getting of the input is then not blocked when new event loop is started (modal screen) but delivering of this message is. Which is demanded behavior.

[the remark this answers is not in the document]
Comment (poncovka) [reply simpleline/input/__init__.py]: Ok.

File: simpleline/input/input_threading.py [hunk 14]
```diff
@@ -0,0 +1,237 @@
+#
```
Comment (poncovka) [inline simpleline/input/input_threading.py]: Please, remove the extra line.

File: simpleline/input/input_threading.py [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+
+    This manager helps with concurrent user input (still you really shouldn't do that).
+    """
+
+    __instance = None
+
+    def __init__(self):
+        super().__init__()
+        self._input_stack = []
+        self._registered_loop = App.get_event_loop()
```
Comment (poncovka) [inline simpleline/input/input_threading.py]: This line can be probably removed.

File: simpleline/input/input_threading.py [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+        self._processing_input = False
+
+    @classmethod
+    def new_instance(cls):
```
Comment (poncovka) [inline simpleline/input/input_threading.py]: Maybe `create_new_instance`?

File: simpleline/input/input_threading.py (hunk 14 above)
Comment (jkonecny12) [reply simpleline/input/input_threading.py]: Add description of the file instead.

File: simpleline/input/input_threading.py (hunk 15 above)
Comment (jkonecny12) [reply simpleline/input/input_threading.py]: Good catch.

File: simpleline/input/input_threading.py (hunk 16 above)
Comment (jkonecny12) [reply simpleline/input/input_threading.py]: Done.

**Comment by jkonecny12** [comment]
**UPDATED:**
* Fixed other suggestions from @poncovka .

**Pull request merged**